### ISSUE TITLE
3.0.0alpha15: crypto/bn/asm/rsaz-avx512.pl needs Perl 5.14

### ISSUE THREAD

**Issue by rainerjung** [openssl/openssl#14990]
The new script crypto/bn/asm/rsaz-avx512.pl uses the non-distructive modifier "/r" for the substitution operator "s" in two places. This feature was introduced in Perl 5.14, but OpenSSL 3.0.0 is expected to compile using Perl 5.10. Two platforms which suffer from this problem are eg. RHEL 6 and SLES 11.

The following trivial fix seems to work for me:

```
--- crypto/bn/asm/rsaz-avx512.pl.orig   2021-04-22 16:33:53.140939443 +0200
+++ crypto/bn/asm/rsaz-avx512.pl        2021-04-22 18:13:28.777239513 +0200
@@ -128,7 +128,8 @@
 #                of data for corresponding AMM operation;
 # _b_offset    - offset in the |b| array pointing to the next qword digit;
 my ($_data_offset,$_b_offset,$_acc,$_R0,$_R0h,$_R1,$_R1h,$_R2,$_k0) = @_;
-my $_R0_xmm = $_R0 =~ s/%y/%x/r;
+my $_R0_xmm = $_R0;
+$_R0_xmm =~ s/%y/%x/;
 $code.=<<___;
     movq    $_b_offset($b_ptr), %r13             # b[i]

@@ -525,7 +526,8 @@
 my ($out,$red_tbl,$red_tbl_idx,$tbl_idx) = @_6_args_universal_ABI;

 my ($t0,$t1,$t2,$t3,$t4) = map("%ymm$_", (0..4));
-my $t4xmm = $t4 =~ s/%y/%x/r;
+my $t4xmm = $t4;
+$t4xmm =~ s/%y/%x/;
 my ($tmp0,$tmp1,$tmp2,$tmp3,$tmp4) = map("%ymm$_", (16..20));
 my ($cur_idx,$idx,$ones) = map("%ymm$_", (21..23));
```

Thanks and regards,

Rainer

**Comment by rainerjung** [comment]
I should have also mentioned the original error (recreated here with "perl -wc"):

```
Bareword found where operator expected at crypto/bn/asm/rsaz-avx512.pl line 131, near "s/%y/%x/r"
Unquoted string "r" may clash with future reserved word at crypto/bn/asm/rsaz-avx512.pl line 131.
Bareword found where operator expected at crypto/bn/asm/rsaz-avx512.pl line 528, near "s/%y/%x/r"
Unquoted string "r" may clash with future reserved word at crypto/bn/asm/rsaz-avx512.pl line 528.
syntax error at crypto/bn/asm/rsaz-avx512.pl line 131, near "s/%y/%x/r"
syntax error at crypto/bn/asm/rsaz-avx512.pl line 528, near "s/%y/%x/r"
crypto/bn/asm/rsaz-avx512.pl had compilation errors.

```

**Comment by t8m** [comment]
Would you like to open a PR with this fix?

**Comment by levitte** [comment]
Huh, I didn't notice that one...

In other places, we've done something like this, which is equivalent:

``` perl
(my $_R0_xmm = $_R0) =~ s/%y/%x/;
```

**Comment by rainerjung** [comment]
If you could apply the trivial fix (I don't care which style) without me opening a PR, that would be great.

**Comment by rainerjung** [comment]
I think you can close this issue. It should be fixed by #15379. The underlying issue #15378 is a duplicate of this one here.

**Comment by t8m** [comment]
Closing as fixed on master.
